### DISCUSSION TITLE
blog/react-query-and-react-context

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#97] [Announcements]
# blog/react-query-and-react-context

Can it make sense to combine React Query with React Context ? Yes, sometimes ...

https://tkdodo.eu/blog/react-query-and-react-context#inject-comments

**Comment by jomefavourite** [comment]
Thanks for always keeping us updated on how to make good decisions when using React Query.

**Comment by Arcades2** [comment]
So cool as always ! Thx

**Comment by mk-nickyang** [comment]
Thanks for another great article!

I think Context is also a good solution when we have multiple  `useQuery` instances mounted in different render cycles, and we only want the parent query to do the fetch job, the rest of query instances should just be reading from the state.

Like in the example below, TQ cannot deduplicate network fetches because Child components are conditionally rendered
```
const Child1 = () => {
    const { data } = useQuery();
    return ...
}

const Child2 = () => {
    const { data } = useQuery();
    return ...
}

const ParentScreen = () => {
    const { isLoading, isError } = useQuery(...);

    if (isLoading) return <LoadingView />

    if (isError) return <ErrorView />

    return (
        <>
            <Child1 />
            <Child2 />
        </>
    )
}
```
But with Context, we can avoid having duplicate network requests + props drilling, and Typescript will be pleased.

The only downside I can think of is, if we have `select` for some `useQuery` instances, there is no out-of-box way for Context to do partial subscribe (unless we use lib like [`use-context-selector`](https://github.com/dai-shi/use-context-selector)).

 Just wondering in this case, do you think if Context is still a better approach OR just setting a `staleTime` as you mentioned [here](https://tkdodo.eu/blog/react-query-as-a-state-manager#customize-staletime)?

**Reply by TkDodo** [reply]
if you just want to avoid the refetch because a new component mounted, setting staleTime is usually enough.

**Comment by p6l-richard** [comment]
Good write up (as usual). 

Hard to think of any other use case than authentication or RBAC. 

The “drawback” becomes a feature in these instances:
Blocking unnecessary outgoing requests to the server by introducing the waterfall. 

Do other examples come to mind?

**Reply by TkDodo** [reply]
we're using this pattern for mandatory data that we need on different pages. But yes, it's mostly user data :)

**Comment by AjayTheWizard** [comment]
I found a Typo in Code Example under the Heading **Being self-contained**.
Here fixed code,
```ts
export const useUserQuery = (id: number) => {
  return useQuery({
    queryKey: ['user', id],
    queryFn: () => fetchUserById(id),
  })
}
export const useCurrentUserQuery = () => {
  const id = useCurrentUserId()

  return useUserQuery(id) // <-- Here's the Fix
}
```

**Reply by TkDodo** [reply]
thanks - can you PR this? code is open source: https://github.com/tkdodo/blog

**Comment by baeharam** [comment]
Thanks for your great article! 👏

**Comment by lgenzelis** [comment]
Great advice, thanks a lot for sharing this Tk!! 😁😁

**Comment by raphaelbadia** [comment]
can't wait for v5!

**Comment by jtgi** [comment]
Nice. Glad it's not just me.
As you mentioned, I don't like side stepping the compiler. I find others may not understand why and I prefer the explicit dependency.

I've been using it with react-router's outlet context and nested routes, works decent, e.g:
```ts
function Routes() {
  return (
    <Routes>
      <Route match="/:id" component={ResourceLayout}>
        <Route match="/edit" ... />
      </Route>
    </Routes>
  )
}

function ResourceLayout() {
  const id = useParams();
  const resourceQuery = useResourceQuery(id);
  
  if (resourceQuery.data) {
    return <Outlet context={resourceQuery.data}
  }

  // [...]
}

function useResourceContext() {
  const context = useOutletContext<ResourceContext>();
  invariant(/* ... */);
  
  return context;
}
```

**Comment by Tiberriver256** [comment]
Nice! I also use React Context as a DI mechanism for React Query. Instead of injecting the data though I inject a service implementation. Makes it easy to swap everything out with mocks as needed.

https://github.com/Tiberriver256/react-query-dependency-injection/tree/3--react-query-dependency-injection/src

**Comment by amin-da** [comment]
thanks for great article
can you explain more about the dependency injection ? what does it mean in programming? what is the difference between it and the  state manager?

**Reply by TkDodo** [reply]
dependency injection is about taking a static value that doesn't change in the lifecycle of your App and passing it to your App. With it, you have to chance to swap out implementations, e.g. for testing.

https://en.wikipedia.org/wiki/Dependency_injection

**Comment by danbeo95** [comment]
Great!
Wait for next sharing

**Comment by sebastianlp** [comment]
Thanks for this, it's really amazing and I'm glad that is a tip since I've been using this to get application settings and I always wonder if it's a bit out of the "rule", I have a question tho, since I found myself doing this when testing components that are wrapped in this provider and bc the provider is in charge of the query, it returns a loader, an error or the children depending on the query status I've always wait until the loader component is not in the screen anymore but I feel it's a repetitive solution, what do you think would be the best approach to test components that are children of this component to avoid waiting for the loader to disappear? I was thinking to provide a customQueryHook that always returns false to `isLoading` & `isError` and returns a mocked `data` object

**Reply by TkDodo** [reply]
> what do you think would be the best approach to test components that are children of this component

you wrap them in a provider that contains data in a non-loading state.

**Comment by OBats** [comment]
Hi Dominik. Top-notch content as always!

Just curious, should we check if the data was successfully fetched before passing it as a value? TypeScript still infers it as possibly undefined:

```
  return (
    <CurrentUserContext.Provider value={currentUserQuery.data}> // data is User | undefined
      {children}
    </CurrentUserContext.Provider>
  )
```

So, would it be a correct approach to do something like this?
```
  return (
    <CurrentUserContext.Provider value={currentUserQuery.isSuccess ? currentUserQuery.data : null}>
      {children}
    </CurrentUserContext.Provider>
  )
```

**Reply by TkDodo** [reply]
My example does checks upfront. By excluding loading (ok, pending in v5) and error, it will always be defined:

```
  if (currentUserQuery.isLoading) {
    return <SkeletonLoader />
  }

  if (currentUserQuery.isError) {
    return <ErrorMessage error={currentUserQuery.error} />
  }

  return (
    <CurrentUserContext.Provider value={currentUserQuery.data}>
      {children}
    </CurrentUserContext.Provider>
  )
```

**Reply by OBats** [reply]
Oh.. now it is clear... I was using `isLoading` with v5.
Thanks)

**Comment by ssrihari4102** [comment]
Thanks for documenting this pattern! It was really helpful when setting up my app since it allowed me to have a top level loading state while ensuring data was defined in each component. I was wondering if there might be a way to use this context approach while also reaping some of the benefits of [tracked queries](https://tkdodo.eu/blog/react-query-render-optimizations#tracked-queries)? I'm struggling to think of a way to do so.

As an example, let's say the data I am fetching is an object with 20 keys. Each of my components uses 1-3 keys from the data object. Right now I'm passing the entire object through context after ensuring there is no loading or error state, so when I update the data, every component re-renders rather than just the components that use the changed fields.

I'd imagine I should either not worry about optimizing renders in this case or somehow split up the context, but was curious if there might still be a way to take advantage of some of the tools react query provides out of the box to assist with such optimizations.

**Reply by TkDodo** [reply]
I think this is just the drawback of context here. once the proposal of using context inside `useMemo` (with the help of `use(Context)`) ships, this should automatically work.

**Reply by ssrihari4102** [reply]
Oh that's interesting. I realized after sending my message but Suspense should help with this issue as well right? Namely, like your blog post mentions, we get a similar architecture like we do with context, but are able to just use react query's API, so I imagine we'll be able to use tracked queries by default

**Comment by AayushKarki714** [comment]
Thanks for the awesome article. Since your are using context for sharing the user data, won't it cause any re-rendering issues if your have large component tree?

**Reply by TkDodo** [reply]
It will only re-render where you use that data. You'll lose the ability to create fine-grained subscriptions with `select` though. I think suspense is generally better at this.

**Comment by caillou** [comment]
Sorry, I am new to all of this. When I follow this pattern, `currentUserQuery.data` is of type `undefined | User` when rendering the `AppInfoContext.Provider`.

```typescript
export const CurrentUserContextProvider = ({
  children,
}: {
  children: React.ReactNode
}) => {
  const currentUserQuery = useCurrentUserQuery()

  if (currentUserQuery.isLoading) {
    return <SkeletonLoader />
  }

  if (currentUserQuery.isError) {
    return <ErrorMessage error={currentUserQuery.error} />
  }

  return ( // currentUserQuery.data is undefined | User
    <CurrentUserContext.Provider value={currentUserQuery.data}>
      {children}
    </CurrentUserContext.Provider>
  )
} 
```

It seems that I need to replace the `isLoading` predicate with `isPending`, and then the type is `User`:

```typescript
export const CurrentUserContextProvider = ({
  children,
}: {
  children: React.ReactNode
}) => {
  const currentUserQuery = useCurrentUserQuery()

  if (currentUserQuery.isPending) {
    return <SkeletonLoader />
  }

  if (currentUserQuery.isError) {
    return <ErrorMessage error={currentUserQuery.error} />
  }

  return ( // currentUserQuery.data is User
    <CurrentUserContext.Provider value={currentUserQuery.data}>
      {children}
    </CurrentUserContext.Provider>
  )
} 
```

**Reply by TkDodo** [reply]
yes, this was a change in v5. See this comment: https://github.com/TkDodo/blog-comments/discussions/97#discussioncomment-8321054

**Comment by yeunoia** [comment]
If you’re using v5, do you prefer using ‎`useSuspenseQuery` instead of context pattern?

also, haven't really thought using context as dependency injection. thanks for the insights!

**Reply by TkDodo** [reply]
Yeah I really like suspense for it, especially when combined with route loaders. see:

- https://tkdodo.eu/blog/tan-stack-router-and-query
